### PR TITLE
feat: Allow printing listen address on startup

### PR DESCRIPTION
**Pull request by flub**
The iroh services can be configured to listen on a random port by using a port number of 0.  This allows printing it on stdout at startup so that you know how to contact the service.

This requires https://github.com/n0-computer/quic-rpc/pull/21 before it can be merged.

# Status

ready to be merged

### CODE REVIEW

File: iroh-p2p/src/rpc.rs [hunk 1]
```diff
@@ -570,19 +576,25 @@ async fn dispatch(s: P2pServer, req: P2pRequest, chan: ServerSocket<P2pService>,
...
-    info!("p2p rpc listening on: {}", addr);
-    let server = create_server::<P2pService>(addr).await?;
-    loop {
-        match server.accept_one().await {
-            Ok((req, chan)) => {
-                tokio::spawn(dispatch(server.clone(), req, chan, p2p.clone()));
-            }
-            Err(cause) => {
-                tracing::debug!("p2p rpc accept error: {}", cause);
+pub(crate) async fn spawn_server(
```
Comment (ramfox) [inline iroh-p2p/src/rpc.rs:579]: 🔥

File: iroh-rpc-client/src/lib.rs [hunk 2]
```diff
@@ -1,20 +1,24 @@
...
 pub use status::{ClientStatus, ServiceStatus, ServiceType, StatusType, HEALTH_POLL_WAIT};
 pub use store::StoreClient;
 
+pub mod client;
+pub mod config;
+pub mod gateway;
+pub mod network;
+pub mod status;
+pub mod store;
+
```
Comment (ramfox) [inline iroh-rpc-client/src/lib.rs:21]: Was this moved for a style-guide reason?

**Review by ramfox — APPROVED** [review]

File: iroh-rpc-client/src/lib.rs (hunk 2 above)
Comment (flub) [reply iroh-rpc-client/src/lib.rs:21]: Ugh, I kind of did that without thinking out of habit.  Seems I'm backed up by the style guide though: https://doc.rust-lang.org/stable/style-guide/items.html#items

> We recommend that imports come before module declarations;

Happy to revert though - IMHO this kind of thing can do whatever unless we configure rustfmt to be strict (it doesn't do this - yet)

Comment (ramfox) [reply iroh-rpc-client/src/lib.rs:21]: Nah, no need to change, I was genuinely just curious. Thanks!

File: iroh-util/src/config.rs [hunk 3]
```diff
@@ -0,0 +1,287 @@
+use std::collections::HashMap;
```
Comment (flub) [inline iroh-util/src/config.rs:1]: I moved all config-related things from iroh-util into this new module.

File: iroh-util/src/config.rs [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+
+    let cfg = builder.build()?;
+    debug!("make_config:\n{:#?}\n", cfg);
+    let cfg: T = cfg.try_deserialize()?;
+    Ok(cfg)
+}
+
+/// Configuration specific to an iroh server processes.
+#[derive(Debug, Clone, Default, Deserialize, Serialize, PartialEq)]
+pub struct ServerConfig {
```
Comment (flub) [inline iroh-util/src/config.rs:194]: This is the new thing here.

**Comment by flub** [comment]
PTAL, this can now be merged as quic-rpc has been updated

**Comment by ramfox** [comment]
Hello! The code that was previously hosted in this repository has been moved to [n0-computer/beetle](https://github.com/n0-computer/beetle). `beetle` is in maintenance mode, but will still accept PRs.

 If you are still interested in getting your PR merged, please re-open your PR on [n0-computer/beetle](https://github.com/n0-computer/beetle).

Check out our blog post for more info on our new direction for `iroh`: https://n0.computer/blog/a-new-direction-for-iroh/

**Pull request closed without merging**